### ISSUE TITLE
Allow client to take a screenshot

### ISSUE THREAD

**Issue by skorokithakis** [Genymobile/scrcpy#112]
Would it be possible to add a command-line argument to allow the client to take a screenshot? I am writing a short script to capture the screen, but it's latency-sensitive, and the default way of using `screencap` takes two full seconds just to encode to JPG. Anything cutting down on that would be very welcome.

**Comment by rom1v** [comment]
> Would it be possible to add **a command-line argument** to allow the client to take a screenshot?

A command-line argument? So it would take a screenshot and exit?

Or a shortcut, to save the current frame to a file somewhere?

**Comment by skorokithakis** [comment]
As I need to control it programmatically, it would be most convenient for me if it could take a screenshot and exit, yes. Right now I am looking into whether I can write a small client to speak the protocol and connect to the device to grab a single frame, any information you may have would be of great help.

**Comment by rom1v** [comment]
That would probably not be faster than `screencap`:

```bash
adb exec-out screencap -p > screenshot.png
```

Starting the java application through `app_process` is already about 800ms, then the screen would be captured, encoded, transmitted, and saved on the client. If this is faster than `screencap`, then `screencap` must be fixed :)

**Comment by skorokithakis** [comment]
Ah, I thought the server would already be running. Yes, you may be right. `screencap` takes 250ms or so to capture, but 2 sec to convert to png, so it's pretty slow. Feel free to close this ticket, thank you!

**Comment by skorokithakis** [comment]
I didn't know about `exec-out` so I was doing a save-then-pull, `exec-out` only takes a few ms, thank you!

**Comment by rom1v** [comment]
> I thought the server would already be running.

Ah ok :)

To be faster, `scrcpy` could be running, and receive (from whatever channel) a request to take a screenshot, it already has the frame, so it would encode/save it locally (on the computer, which is probably faster than the device to encode a png).

Too specific though.

**Comment by skorokithakis** [comment]
Yeah, too specific and `screencap` is already fast enough with the `exec-out` trick.

Thanks for the help and for a fantastic project!

**Comment by ooing** [comment]
@skorokithakis   but  I need 2 second

**Comment by dedosmedia** [comment]
> To be faster, `scrcpy` could be running, and receive (from whatever channel) a request to take a screenshot, it already has the frame, so it would encode/save it locally (on the computer, which is probably faster than the device to encode a png).
> 

Any hint how could I achieve this? I am interested in to develop such a feature.
I am already saving a bitmap from the SDL window, but the picture quality is not that great, have H.264 artifacts.
What you described on that comment would improve picture quality?

Thanks Romain

**Comment by rom1v** [comment]
@dedosmedia see #684

**Comment by mankidal19** [comment]
> > I thought the server would already be running.
> 
> Ah ok :)
> 
> To be faster, `scrcpy` could be running, and receive (from whatever channel) a request to take a screenshot, it already has the frame, so it would encode/save it locally (on the computer, which is probably faster than the device to encode a png).
> 
> Too specific though.

Dear @rom1v, thank you for your workaround suggestion. Now I managed to write a simple bash script to automatically execute scrcpy, request screenshot in ubuntu using gnome-screenshot (no window), save it, & close the scrcpy window. The problem is the captured image resolution is not correct as my device (an Android TV) has a 2k resolution. This is because my PC monitor aspect ratio is 16:10 while the TV itself aspect ratio is 16:9. (Already using ctrl + G shortcut). However, the tool I'm developing has a requirement to preserve the original device resolution. Any idea how to solve this? adb screencap doesn't always reliable because it can't capture some screens (especially 3rd party apps i.e., YouTube main screen), which will return an empty png file.
